airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ProtoToCsvWriter.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.object_storage
+
+import com.fasterxml.jackson.core.io.JsonStringEncoder
+import io.airbyte.cdk.load.data.*
+import io.airbyte.cdk.load.message.*
+import java.time.Instant
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+
+class ProtoToCsvWriter(
+    private val header: Array<String>,
+    private val columns: Array<AirbyteValueProxy.FieldAccessor>,
+    private val rootLevelFlattening: Boolean,
+    private val extractedAtTsTz: Boolean,
+) {
+
+    private val idxRawId = header.indexOf(Meta.COLUMN_NAME_AB_RAW_ID)
+    private val idxExtractedAt = header.indexOf(Meta.COLUMN_NAME_AB_EXTRACTED_AT)
+    private val idxMeta = header.indexOf(Meta.COLUMN_NAME_AB_META)
+    private val idxGenerationId = header.indexOf(Meta.COLUMN_NAME_AB_GENERATION_ID)
+    private val idxData = header.indexOf(Meta.COLUMN_NAME_DATA)
+    private val jsonEscaper = JsonStringEncoder.getInstance()
+
+    private val rowIndex: Map<String, Int> =
+        if (rootLevelFlattening) {
+            columns.associate { it.name to header.indexOf(it.name) }
+        } else {
+            emptyMap()
+        }
+
+    private val rowBuf: Array<Any?> = arrayOfNulls(header.size)
+
+    fun toCsvRow(
+        record: DestinationRecordRaw,
+        src: DestinationRecordProtobufSource,
+        changes: List<Meta.Change>
+    ): Array<Any?> {
+        rowBuf[idxRawId] = record.airbyteRawId.toString()
+
+        val emitted = src.emittedAtMs
+        rowBuf[idxExtractedAt] =
+            if (extractedAtTsTz)
+                OffsetDateTime.ofInstant(Instant.ofEpochMilli(emitted), ZoneOffset.UTC).toString()
+            else emitted
+
+        rowBuf[idxMeta] = buildMeta(record.stream.syncId, changes)
+        rowBuf[idxGenerationId] = record.stream.generationId
+
+        val proxy = src.asAirbyteValueProxy()
+
+        if (rootLevelFlattening) {
+            var i = 0
+            while (i < columns.size) {
+                val column = columns[i]
+                val index =
+                    rowIndex[column.name]
+                        ?: throw IllegalArgumentException(
+                            "Column '${column.name}' not found in row index"
+                        )
+                rowBuf[index] = fetchValue(proxy, column)
+                i++
+            }
+        } else {
+            // one big JSON blob under _airbyte_data
+            rowBuf[idxData] = buildPayloadJson(proxy)
+        }
+
+        return rowBuf
+    }
+
+    private fun fetchValue(
+        proxy: AirbyteValueProxy,
+        field: AirbyteValueProxy.FieldAccessor,
+    ): Any {
+        return when (field.type) {
```

Comment:
checking my understanding: in protobuf, we're guaranteed that `field.type` matches the actual type on the record, so we don't need to do anything to guard against mismatches here?

Reply:
Yup